source/extensions/filters/http/cache/cache_filter.h

```diff
@@ -38,19 +39,50 @@ class CacheFilter : public Http::PassThroughFilter,
   Http::FilterDataStatus encodeData(Buffer::Instance& buffer, bool end_stream) override;
 
 private:
+  // Utility functions; make any necessary checks and call the corresponding lookup_ functions
   void getBody();
-  void onHeaders(LookupResult&& result);
+  void getTrailers();
+
+  // Callbacks for HttpCache to call when headers/body/trailers are ready
+  void onHeaders(LookupResult&& result, Http::RequestHeaderMap& request_headers);
   void onBody(Buffer::InstancePtr&& body);
   void onTrailers(Http::ResponseTrailerMapPtr&& trailers);
 
+  // Precondition: lookup_result_ points to a cache lookup result that requires validation
+  //               filter_state_ is ValidatingCachedResponse
+  // Serves a validated cached response after updating it with a 304 response
+  void processSuccessfulValidation(Http::ResponseHeaderMap& response_headers);
+
+  // Precondition: lookup_result_ points to a cache lookup result that requires validation
+  //               filter_state_ is ValidatingCachedResponse
+  // Checks if a cached entry should be updated with a 304 response
+  bool shouldUpdateCachedEntry(const Http::ResponseHeaderMap& response_headers) const;
+
+  // Precondition: lookup_result_ points to a cache lookup result that requires validation
+  // Should only be called during onHeaders as it modifies RequestHeaderMap
+  // Adds required conditional headers for cache validation to the request headers
+  // according to the present cache lookup result headers
+  void injectValidationHeaders(Http::RequestHeaderMap& request_headers);
+
+  // Precondition: lookup_result_ points to a fresh or validated cache look up result
+  //               filter_state_ is ValidatingCachedResponse
+  // Adds a cache lookup result to the response encoding stream
+  // Can be called during decoding if a valid cache hit is found
+  // or during encoding if a cache entry was validated successfully
+  void encodeCachedResponse();
+
+  // Precondition: finished adding a response from cache to the response encoding stream
+  // Updates the filter_state_ and continues the encoding stream if necessary
+  void finishedEncodingCachedResponse();
+
   TimeSource& time_source_;
   HttpCache& cache_;
   LookupContextPtr lookup_;
   InsertContextPtr insert_;
+  LookupResultPtr lookup_result_;
```

Comment:
We should either make lookup_result_ be a plain LookupResults (replacing !lookup_result with , or we should change makeLookupResult to return a LookupResultPtr. I lean towards the former (to avoid the allocation/indirection), but either would be ok.